plugin/storage/cassandra/spanstore/reader.go

```diff
@@ -177,26 +238,135 @@ func (s *SpanReader) FindTraces(queryParams *spanstore.TraceQueryParameters) ([]
 	return retMe, nil
 }
 
-// executeQuery returns UniqueTraceIDs
-func (s *SpanReader) executeQuery(query Query, tableMetrics *casMetrics.Table) (dbmodel.UniqueTraceIDs, error) {
+// FindTraces queries for traces.
+func (s *SpanReader) getTraceIDs(traceQuery *spanstore.TraceQueryParameters) ([]dbmodel.TraceID, error) {
+	if traceQuery.DurationMin != 0 {
+		return s.queryByDuration(traceQuery)
+	}
+
+	// NB(rooz): consider refactoring the query... methods where they take in traceQuery instead of a breakdown of its values
+	if traceQuery.OperationName != "" {
+		traceIds, err := s.queryByServiceNameAndOperation(traceQuery.ServiceName, traceQuery.OperationName, traceQuery.StartTimeMin, traceQuery.StartTimeMax, traceQuery.NumTraces)
+		if err != nil {
+			return nil, err
+		}
+		if len(traceQuery.Tags) > 0 {
+			tagTraceIds, err := s.queryByTagsAndLogs(traceQuery.ServiceName, traceQuery.StartTimeMin, traceQuery.StartTimeMax, traceQuery.Tags, traceQuery.NumTraces)
+			if err != nil {
+				return nil, err
+			}
+			return s.intersectTraceIDSlices(traceIds, tagTraceIds), nil
+		}
+		return traceIds, err
+	}
+	if len(traceQuery.Tags) > 0 {
+		return s.queryByTagsAndLogs(traceQuery.ServiceName, traceQuery.StartTimeMin, traceQuery.StartTimeMax, traceQuery.Tags, traceQuery.NumTraces)
+	}
+	return s.queryByService(traceQuery.ServiceName, traceQuery.StartTimeMin, traceQuery.StartTimeMax, traceQuery.NumTraces)
+}
+
+func (s *SpanReader) intersectTraceIDSlices(one, other []dbmodel.TraceID) []dbmodel.TraceID {
+	oneAsMap := dbmodel.UniqueTraceIDs{}
+	for _, i := range one {
+		oneAsMap[i] = struct{}{}
+	}
+	var retMe []dbmodel.TraceID
+	for _, i := range other {
+		if _, ok := oneAsMap[i]; ok {
+			retMe = append(retMe, i)
+		}
+	}
+	return retMe
+}
+
+func (s *SpanReader) queryByTagsAndLogs(serviceName string, startTime, endTime time.Time, tags map[string]string, limit int) ([]dbmodel.TraceID, error) {
+	results := dbmodel.UniqueTraceIDs{}
+	for k, v := range tags {
+		// service_name = ? AND tag_key = ? AND tag_value = ? and start_time > ? and start_time < ? limit ?`
+		query := s.session.Query(
+			queryByTag,
+			serviceName,
+			k,
+			v,
+			model.TimeAsEpochMicroseconds(startTime),
+			model.TimeAsEpochMicroseconds(endTime),
+			limit,
+		)
+		t, err := s.executeQuery(query, s.readerMetrics.queryTagIndex)
+		if err != nil {
+			return nil, err
+		}
+		for _, tID := range t {
+			results.Add(tID)
+		}
+	}
+	return results.ToList(), nil
```

Comment:
no reason to convert to list here, since at the very top FindTraces function you still go back to unique IDs. I think all of query functions should simply operate on UniqueIDs